Particle sampler implementation

denainjs opened the pull request:
Library for particle samplers in Turing.

devmotion commented on Project.toml:
```diff
@@ -0,0 +1,14 @@
+name = "AdvancedPS"
+uuid = "9c00393c-822d-4781-8df5-4c3c33f9866d"
+authors = ["JS Denain <denain.js@gmail.com>"]
+version = "0.1.0"
+
+[deps]
+AbstractMCMC = "80f14c24-f653-4e6a-9b94-39d6b0f70001"
+Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
+DynamicPPL = "366bfd00-2699-11ea-058f-f148b4cae6d8"
+Libtask = "6f1fad26-d15e-5dc8-ae53-837a1d7b8c9f"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+StatsFuns = "4c63d2b9-4356-54db-8cca-17b64c39e42c"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+Turing = "fce5fe82-541a-59a6-adf8-730c64b5f9a0"
```
The Manifest.toml file should be removed and Project.toml should be cleaned: test dependencies have to be moved to a separate section (e.g., this package shouldn't depend on Test or Turing, and probably also not on DynamicPPL), and compatibility bounds have to be specified.

denainjs replied:
what do you mean by 'moved to a separated section'?

devmotion replied:
Yes, exactly :+1:

devmotion commented on README.md:
```diff
@@ -1,23 +1 @@
-
-
-
+# AdvancedPS.jl
```
Why are the references removed?

denainjs replied:
my bad, changed

devmotion commented on src/particlecontainer.jl:
```diff
@@ -0,0 +1,85 @@
+end
+
+"""
+    reset_logweights!(pc::ParticleContainer)
```
Copying messed up the docstrings.

devmotion commented on test/resampling.jl:
```diff
@@ -0,0 +1,17 @@
+    resResidual   = resample_residual(D, num_samples )
+
+
+    @test sum(resSystematic .== 2) ≈ (num_samples * 0.4) atol=1e-3*num_samples
+    @test sum(resStratified .== 2) ≈ (num_samples * 0.4) atol=1e-3*num_samples
+    @test sum(resMultinomial .== 2) ≈ (num_samples * 0.4) atol=1e-2*num_samples
+    @test sum(resResidual .== 2) ≈ (num_samples * 0.4) atol=1e-2*num_samples
```
```suggestion
    @test count(==(2), resSystematic) ≈ 0.4 * num_samples 0.4atol=1e-3*num_samples
    @test count(==(2), resStratified) ≈ 0.4 * num_samples atol=1e-3*num_samples
    @test count(==(2), resMultinomial) ≈ 0.4 * num_samples atol=1e-2*num_samples
    @test count(==(2), resResidual) ≈ 0.4 * num_samples atol=1e-2*num_samples
```
Why are the different tolerances needed BTW?

cpfiffer commented:
I share @devmotion's concern that it's hard to review this. Could we basically just copy all the Turing code first in a separate PR, merge that through first, and then use the new master to compare the changes made as part of this PR?

denainjs commented:
AdvancedPS here is supposed to work with the Turing samplers introduced in [this PR](https://github.com/TuringLang/Turing.jl/pull/1381).
I don't think it addresses the issues in [TuringLang/Turing.jl#1345](TuringLang/Turing.jl#1345).

cpfiffer commented:
Yeah, that's fine -- it's just difficult to examine the changes because all the code here is "new" to AdvancedPS.jl, so we can't review changes that might be a bad idea because there's nothing to compare to.

yebai commented:
Close in favour of #14
